### ISSUE TITLE
Matching algorithm in rewriting does not reduce any pattern-matching definitions

### ISSUE THREAD

**Issue by L-TChen** [agda/agda#7654]
The matching algorithm used in rewriting does not apply the pattern substitution during matching to indices, even though the pattern variable has been matched with a term. 

Consider the following pathological example 
```agda
{-# OPTIONS --rewriting -v rewriting.match:50 #-}

open import Agda.Builtin.Equality.Rewrite
open import Agda.Builtin.Equality
open import Agda.Builtin.Nat

data T : Nat → Set where
  U : (n : Nat) → T n
  
postulate 
  f  : (n : Nat) → T (n + 0) → T n
  fn : (n : Nat) → f n (U (n + 0)) ≡ U n

  {-# REWRITE fn #-}
  
f0 : f 0 (U 0) ≡ U 0
f0 = fn 0 -- refl should suffice!
```
where `f 0 (U 0)` should reduce to `U 0` since `f 0` is of type `T (0 + 0) → T 0` and `U 0` is equal to `U (0 + 0)`.

### Agda debug 
According to the following excerpt from `*Agda debug*`, the rule `fn` is indeed considered and and the pattern variable `n` has been matched with `0`. However, the pattern variable occurring in `U (n + 0)` is not substituted with `0`, causing the later mismatch between `n + 0` and `0`. 
```
{ attempting to rewrite term  f zero (U zero)
 having head  f  of type  (n : Nat) → T (n + 0) → T n
 with rule  fn (n : Nat)  |- 
            (f n (U (_+_ n .(0))))  -->  U n  :  T n
matching elimination  n
  with                $ zero
  eliminating head    f : (n : Nat) → T (n + 0) → T n
matching pattern  n   with term       0   of type         Nat
matching a PVar:  0
matching elimination  (U (_+_ n .(0)))
  with                $ U zero
  eliminating head    f zero : T (zero + 0) → T zero
matching pattern  (U (_+_ n .(0)))
  with term       U zero
  of type         T (zero + 0)
matching a PDef:  U
matching elimination  (_+_ n .(0))
  with                $ zero
  eliminating head    U : (n : Nat) → T n
matching pattern  (_+_ n .(0))
  with term       0
  of type         Nat
matching a PDef:  _+_
mismatch between (_+_ n .(0))  and  0  of type  Nat 
blocking tag from reduction:  NotBlocked {blockingStatus = ReallyNotBlocked, ignoreBlocking = ()}
matching failed during matching
blocking:  NotBlocked {blockingStatus = ReallyNotBlocked, ignoreBlocking = ()}
}
failed to rewrite  f zero (U zero)
blocking tag NotBlocked {blockingStatus = ReallyNotBlocked, ignoreBlocking = ()}
```

I have tested the above example with `Agda 2.8.0-e8ab708`.

Is this a bug? Or, do I overlook anything?

**Comment by jespercockx** [comment]
Your rewrite rule breaks confluence, with `--confluence-check` you get two warnings:
```
Global confluence check failed: f (suc n) (U (suc n + 0)) can be
rewritten to either U (suc n) or f (suc n) (U (suc (n + 0))).
Possible fix: add a rewrite rule with left-hand side
f (suc n) (U (suc n + 0)) to resolve the ambiguity.
when checking confluence of the rewrite rule fn with
Agda.Builtin.Nat._+_-clause2
```
and
```
Global confluence check failed: f zero (U (zero + 0)) can be
rewritten to either U zero or f zero (U 0).
Possible fix: add a rewrite rule with left-hand side
f zero (U (zero + 0)) to resolve the ambiguity.
when checking confluence of the rewrite rule fn with
Agda.Builtin.Nat._+_-clause1
```
In this case the suggestion that Agda makes is more complex than necessary, it suffices to make `_+_` compute on both arguments:
```agda
postulate
  +zero : (m : Nat) → m + 0 ≡ m
  +suc : (m n : Nat) → m + (suc n) ≡ suc (m + n)
  +suczero : (m : Nat) → suc m + 0 ≡ suc m
  +zerosuc : (m : Nat) → 0 + suc m ≡ suc m
  +sucsuc : (m n : Nat) → (suc m) + (suc n) ≡ suc (suc (m + n))

{-# REWRITE +zero +suc +suczero +zerosuc +sucsuc #-}
```
(The latter three rules are for helping the confluence checker resolve the critical pairs.)

With these rules added, the proof `f0 = refl` works.

Moral of the story: if it seems like a rewrite rule fails to fire, always first check if your rules are confluent (or at least locally confluent, with `--local-confluence-check`)!

**Comment by L-TChen** [comment]
I was aware that this definition does not pass the (local) confluence checker and adding a symmetric addition would solve this. 

The problem I’m confused with is in this pathological example after this matching
```
matching pattern  n   with term       0   of type         Nat
matching a PVar:  0
```
 the pattern variable `n` should have been unified with `0` but it still occurs as `n` in later matching problems, e.g., 
```
matching elimination  (U (_+_ n .(0)))
  with                $ U zero
  eliminating head    f zero : T (zero + 0) → T zero
matching pattern  (U (_+_ n .(0)))
  with term       U zero
  of type         T (zero + 0)

```

**Comment by jespercockx** [comment]
Ah, this is just how the pattern matching for rewrite rules is implemented: it first matches all parts of the pattern independently (i.e. without propagating any information about variable assignments) and after that it does a check that the variable assignments are consistent.

Even if we did do this substitution more eagerly, it still wouldn't help in your example because patterns do not reduce. I.e. you would end up with the pattern `_+_ .(0) .(0)` which would still fail to match against `zero`.

**Comment by L-TChen** [comment]
> Ah, this is just how the pattern matching for rewrite rules is implemented: it first matches all parts of the pattern independently (i.e. without propagating any information about variable assignments) and after that it does a check that the variable assignments are consistent.
> 
> Even if we did do this substitution more eagerly, it still wouldn't help in your example because patterns do not reduce. I.e. you would end up with the pattern `_+_ .(0) .(0)` which would still fail to match against `zero`.

Hum, the above postulated function should be equivalent to the following function `g`
```agda
g : (n : Nat) → T (n + 0) → T (n + 0)
g n (U .(n + 0)) = U (n + 0)
```
where `g zero (U zero)` does compute:
```agda
gn : g zero (U zero) ≡ U zero
gn = refl
```

Given that the doc says 
> More precisely, a term (definitionally equal to) `f p₁σ ... pₙσ` is rewritten to `vσ`, where σ is any substitution on the pattern variables x₁, … xₖ.

I was expecting Agda uses a more flexible matching algorithm to find the unifier, such as turning variables in the local context into metavariables and apply the existing unification algorithm with the stuck term in question.

(I didn't find this design choice in the RTT paper. If I miss the reason for this, please let me know. Thanks!)

**Comment by jespercockx** [comment]
> Hum, the above postulated function should be equivalent to the following function

Unlike dot patterns in regular dependent pattern matching, terms embedded in rewrite rule patterns cannot in general be assumed to match based on typing, so they are instead checked. So it is not the case that turning a clause of a definition into a rewrite rule gives equivalent computational behavior in general.

> I was expecting Agda uses a more flexible matching algorithm to find the unifier, such as turning variables in the local context into metavariables and apply the existing unification algorithm with the stuck term in question.

There's definitely ways to fix this (either by detecting forced arguments like in pattern matching, or by switching from matching to unification as you suggest) but I cannot help but wonder why you would not just define confluent rules to begin with (in which case there is no difference between the algorithms).

> (I didn't find this design choice in the RTT paper. If I miss the reason for this, please let me know. Thanks!)

The matching algorithm used for rewrite rules is described in detail in my paper "Type Theory Unchained": https://drops.dagstuhl.de/entities/document/10.4230/LIPIcs.TYPES.2019.2

**Comment by L-TChen** [comment]
> but I cannot help but wonder why you would not just define confluent rules to begin with (in which case there is no difference between the algorithms).

Because I didn’t find one! 😵
The definition I am working on (instead of the above pathological one) would be locally confluent if the matching algorithm was unification (or something more computational), so I was totally puzzled. Anyway, now I know this is not unintended, and changing matching to unification is also a possibility (at least not something known to fail). 

If interested, the following excerpt is from the definition I am actually working on where `U[]` (i.e. the type substitution with liftings for `U`) is not locally confluent because it does not fire for the special cases `U[∅]` and `U[,]`. (I have tried replacing the index of `Ty (Δ ++ As)` in `subTy` with a variable and a separate identity proof, but then some other rules fail to be proper rewrite rules.)

```agda
{-# OPTIONS --local-confluence-check #-}
open import Agda.Builtin.Equality
open import Agda.Builtin.Equality.Rewrite
open import Agda.Primitive

infixl 10 _++_

interleaved mutual
  data Ctx : Set
  data Lift : Ctx → Set
  data Ty  : Ctx → Set
  data Sub : Ctx → Ctx → Set
  data Tm  : (Γ : Ctx) → Ty Γ → Set

  variable
      Γ Δ Θ Φ   : Ctx

  _++_ : (Γ : Ctx) → Lift Γ → Ctx

  postulate
    _[_]l : Lift Δ → Sub Γ Δ → Lift Γ
    subTy : (As : Lift Δ) (σ : Sub Γ Δ) (A : Ty (Δ ++ As)) → Ty (Γ ++ (As [ σ ]l))
  
  syntax subTy As σ A = A [ σ ⇈ As ]
  
  data _ where
    ∅   : Ctx
    _,_ : (Γ : Ctx) (A : Ty Γ) → Ctx
    ∅   : Lift Γ
    _,_ : (As : Lift Γ)(A : Ty (Γ ++ As)) → Lift Γ
    U   : Ty Γ

  Γ ++ ∅        = Γ
  Γ ++ (As , A) = (Γ ++ As) , A

  postulate
    U[]  : (Γ Δ : Ctx) (σ : Sub Γ Δ) {As : Lift Δ} → U [ σ ⇈ As ] ≡ U

  U[∅]  : (Γ Δ : Ctx) (σ : Sub Γ Δ) → subTy ∅ σ U ≡ U
  U[∅] _ _ _ = U[] _ _ _
  U[,]  : (Γ Δ : Ctx) (σ : Sub Γ Δ) {As : Lift Δ} {A : Ty (Δ ++ As)} → subTy (As , A) σ U ≡ U
  U[,] _ _ _ = U[] _ _ _

  {-# REWRITE U[]  #-}
--  {-# REWRITE U[∅] U[,] #-} -- Adding these two rules does not help at all.
```

**Comment by jespercockx** [comment]
I see, thank you for the example. While I see how unification matching could be useful here, it also would take us farther outside the lands explored by rewriting theory. For the current algorithm, we at least have an idea of how to check the rules for confluence, but if we start interleaving matching and reduction then it feels like really all bets are off.

**Comment by L-TChen** [comment]
Right, thanks for clarifying. For now, I would like to propose to change the wording in the documentation: 

> More precisely, a term (definitionally equal to) f p₁σ ... pₙσ is rewritten to vσ, where σ is any substitution on the pattern variables x₁, … xₖ.

since "a term (definitionally equal to) `f p₁σ ... pₙσ`" suggests that `f 0 (U (0 + 0))`, which is definitionally equal to `f 0 (U 0)`, should fire the rule in the pathological example. The question is, I don't know if there is a particular name for this equality, or the matching algorithm should be presented instead.

**Comment by jespercockx** [comment]
Yeah, I think the wording using definitional equality still stems from the very first experimental implementation of rewriting where we tried to reuse Agda's metavariable solver for the matching (it didn't work very well). Right now the matching algorithm matches things up to beta, eta and irrelevance, but does not reduce any pattern-matching definitions.

**Comment by L-TChen** [comment]
Good. I will file a PR for the documentation in a few days.

**Comment by L-TChen** [comment]
If I change the definitions to the following
```agda
data T' (n : Nat) : Set where
  U : T' n

postulate
  h : (n : Nat) → T' (n + 0) → T' (n + 0)
  hn : (n : Nat) → h n (U {n + 0}) ≡ U {n + 0}

  {-# REWRITE hn #-}
```
Then, the rewrite rule does fire
```agda
h0 : h 0 (U {0}) ≡ U {0}
h0 = refl
```
and passes the confluence checker. The `Agda debug` shows that the unifier `n = 0` is actually applied:
```
{ attempting to rewrite term  h 0 U
 having head  h  of type  (n : Nat) → T' (n + 0) → T' (n + 0)
 with rule  hn (n : Nat)  |-  (h n (U))  -->  U  :  T' (n + 0)
matching elimination  n
  with                $ 0
  eliminating head    h : (n : Nat) → T' (n + 0) → T' (n + 0)
matching pattern  n   with term       0   of type         Nat
matching a PVar:  0
matching elimination  (U)
  with                $ U
  eliminating head    h 0 : T' (0 + 0) → T' (0 + 0)
matching pattern  (U)
  with term       U
  of type         T' (0 + 0)
matching a PDef:  U
rewrote  h 0 U  to  U }
```
IIUC, parameters are not regarded as arguments, so the LHS of this rule is matched successfully. Is that right?

**Comment by andreasabel** [comment]
@jespercockx @L-TChen The progress here is stuck on the last question, it seems...

> parameters are not regarded as arguments, so the LHS of this rule is matched successfully. Is that right?

**Comment by jespercockx** [comment]
Indeed, parameters of constructors and projections (as well as projection-like functions) are not stored in the internal syntax and so are not considered by the matching algorithm; they are assumed to match when all the other arguments do.
